Restrict signature algorithmn via cnf.jwk.alg (#115)
ietf-wg-wimse/draft-ietf-wimse-s2s-protocol: draft-ietf-wimse-s2s-protocol.md

    * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).
      WITs should be refreshed regularly, e.g. on the order of hours.
    * `jti`: A unique identifier for the token. This claim is OPTIONAL. The `jti` claim is frequently useful for auditing issuance of individual WITs or to revoke them, but some token generation environments do not support it.
    * `cnf`: A confirmation claim referencing the public key of the workload.
        * `jwk`: The public key of the workload as defined in {{Section 3.2 of RFC7800}}. The workload MUST prove possession of the corresponding private key when presenting the WIT to another party, which can be accomplished by using it in conjunction with one of the methods in {{dpop-esque-auth}} or {{http-sig-auth}}. As such, it MUST NOT be used as a bearer token and is not intended for use in the `Authorization` header.

>             * `alg`: The signature algorithmn allowed to be used for the public key. The presented proof (WPT or http-sig) MUST be produced with the algorithmn specified in this field.

Author: Updated, please take a look.

    * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).
      WITs should be refreshed regularly, e.g. on the order of hours.
    * `jti`: A unique identifier for the token. This claim is OPTIONAL. The `jti` claim is frequently useful for auditing issuance of individual WITs or to revoke them, but some token generation environments do not support it.
    * `cnf`: A confirmation claim referencing the public key of the workload.
        * `jwk`: Within the cnf claim, a `jwk` key MUST be present that contains the public key of the workload as defined in {{Section 3.2 of RFC7800}}. The workload MUST prove possession of the corresponding private key when presenting the WIT to another party, which can be accomplished by using it in conjunction with one of the methods in {{dpop-esque-auth}} or {{http-sig-auth}}. As such, it MUST NOT be used as a bearer token and is not intended for use in the `Authorization` header.

>             * `alg`: Within the jwk object, an `alg` field MUST be present. Allowed values are listed in the IANA "JSON Web Signature and Encryption Algorithms" registry established by {{RFC7518}}. The presented proof (WPT or http-sig) MUST be produced with the algorithm specified in this field. The value `none` MUST NOT be used. Algorithms used in combination with symmetric keys MUST NOT be used. Also encryption algorithmns MUST NOT be used as this would require additional key distribution outside of the WIT. To promote interoperability, the `ES256` signing algorithm SHOULD be supported by general purpose implementations of this document.

Reviewer: I still think "must" is the right approach for interop. What this means is that general-purpose libraries need to support ES256. Specific implementations are free to use any other algorithms.

Author: I'm ok with that, changed to MUST.

{
  "cnf": {
    "jwk": {

>       "alg": "EdDSA",

Author: Note for myself from SPIFFE sig-spec: This restricts WIT to be used with JOSE only.

Reviewer 2: Per the pointer that @jricher gave in https://github.com/ietf-wg-wimse/draft-ietf-wimse-s2s-protocol/pull/115/files#r2070242131 to  [RFC9421-3.3.7](https://www.rfc-editor.org/rfc/rfc9421.html#name-json-web-signature-jws-algo) - I don't think it actually does restrict the WIT to be used with JOSE only. No?

Reviewer 3: It would effectively restrict it to use JOSE-defined algs but the HTTPSig RFC at least says how to map a JOSE JWS alg to the HTTPSig primitives (`HTTP_SIGN` and `HTTP_VERIFY`) so for that case you're OK. It doesn't necessarily allow for future usages with other presentation mechanisms but I think that's fine.

* identifies the workload to which the token was issued as `wimse://example.com/specific-workload`.
* has a unique identifier of `x-_1CTL2cca3CSE4cwb__`.
* binds the public key represented by the `jwk` confirmation method to the workload `wimse://example.com/specific-workload`.
* requires the proof to be produced with the `EdDSA` signature algorithm.

> For elucidative purposes only, the workload's key, including the private part, is shown below in JWK {{RFC7517}} format:

Author: I don't think the keys in jwk presentation in the next 2 examples represent the workloads public key the way it's included in the WIT. 

Directly after your comment: Workload private key (we could add alg into it, but its irrelevant)
The one following: WIT-signing public key (can be a whole different algorithm all-together, here EcDSA compared to workloads EdDSA)

> For elucidative purposes only, the workload's key, including the private part, is shown below in JWK {{RFC7517}} format:

Reviewer: I agree.

* `nonce`
* `tag` - the value for implementations of this specification is `wimse-workload-to-workload`

The following signature parameters in the `Signature-Input` header MUST NOT be used:

> * `keyid` - The signing key is sent along with the message in the WIT. Additionally specifying the id of the key adds confusion.

Reviewer: ```suggestion
* `keyid` - The signing key is sent along with the message in the WIT. Additionally specifying the key identity would add confusion.
```

> * `alg` - The signature algorithm is specified in the `jwk` section of the `cnf` claim in the WIT. See {{to-wit}} for details.

Reviewer: ```suggestion
* `alg` - The signature algorithm is specified in the `jwk` section of the `cnf` claim in the WIT. See {{to-wit}} and Sec. 3.3.7 of {{RFC9421}} for details.
```

To detect message replays,
a recipient MAY reject a message (request or response) if a nonce is repeated.

<!-- What do we do here? -->

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer: I'm not sure what you're asking: to me this says that implementations must support this specific algorithm. But if the WIT explicitly requests another algorithm, that's totally fine. IOW I'm fine with the text as-is.

As a second order consideration, it would be kinda nice if our examples actually used the mandatory algorithm, but that's not strictly necessary.

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer: I thought it was clear :-(

Anyway, I'm fine if you want to change it to "supported".

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer 3: If there's a reliable place for an algorithm identifier that's outside the signature itself, we should use that. The claim within the WIT makes sense to me, the only question is whether that's a JOSE alg (seems to be) or an HTTPSig Alg (seems not to be). That just needs to be very clear.

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer: Agree with Justin. Within the WIT this would be a JOSE alg, and AFAIK there is no normative mapping from JOSE algs to HTTPSig algs. This is a problem but IMO not a showstopper.

I'm wondering if we should say in Sec. 4.3 that the receiver MUST ignore any incoming `alg` parameter in the signature.

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer 3: There is not a mapping from a JOSE `alg` value to an HTTPSig `alg` value, however there is a normative mapping from a JOSE alg definition to the HTTPSig process, in https://www.rfc-editor.org/rfc/rfc9421.html#name-json-web-signature-jws-algo and I would recommend that we use this explicitly if we can count on their being a JOSE `alg` in a trusted place. GNAP does something similar since it carries the key in its initial request.

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer 2: In this case the alg is coming from the cnf.jwk of the WIT so is trusted as much as anything happening in this context. 

[JWK says its alg](https://www.rfc-editor.org/rfc/rfc7517#section-4.4) should either be registered in the IANA "JSON Web Signature and Encryption Algorithms" registry ... or be a value that contains a Collision-Resistant Name" which kinda suggests you shouldn't use something like ecdsa-p256-sha256 but doesn't outright prohibit it. 

With that normative mapping from JOSE to HTTPSig that @jricher pointed to, it would probably make the most sense for the WIT to always use JOSE alg values in cnf.jwk.alg because that still allows for either the WPT or HTTPsig proof styles to be used.

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Author: @yaronf @jricher @bc-pi 
I am now restricting it to JOSE alg and forbid using the alg and keyid parameters when using http-sig. This aligns with RFC9421 stating in [Section 3.3.7](https://www.rfc-editor.org/rfc/rfc9421.html#section-3.3.7):
> JSON Web Algorithm (JWA) values from the "JSON Web Signature and Encryption Algorithms" registry are not included as signature parameters.

The `keyid` one was previously SHOULD NOT, I personally don't see any reason why to include it and believe it would create uncertainty when it's included so I changed this to MUST NOT. 

What do you think?

> To promote interoperability, the `ecdsa-p256-sha256` signing algorithm MUST be implemented

Reviewer: I'm good with both decisions. Added a review comment to include a reference to Sec. 3.3.7.
